crates/katana/executor/src/blockifier/utils.rs

```diff
@@ -87,6 +99,103 @@ pub fn estimate_fee(
+pub fn simulate_transaction(
+    transaction: ExecutableTxWithHash,
+    block_context: &BlockContext,
+    state: Box<dyn StateProvider>,
+    validate: bool,
+    charge_fee: bool,
+) -> Result<SimulatedTransaction, TransactionExecutionError> {
+    let state = CachedStateWrapper::new(StateRefDb::from(state));
+    let result = TransactionExecutor::new(
+        &state,
+        block_context,
+        charge_fee,
+        validate,
+        vec![transaction.clone()].into_iter(),
+    )
+    .with_error_log()
+    .next()
+    .ok_or(TransactionExecutionError::ExecutionError(
+        EntryPointExecutionError::ExecutionFailed { error_data: Default::default() },
+    ))?;
+    let result = result?;
+
+    let function_invocation =
+        result.execute_call_info.as_ref().map(function_invocation_from_call_info).ok_or(
+            TransactionExecutionError::ExecutionError(EntryPointExecutionError::ExecutionFailed {
+                error_data: Default::default(),
+            }),
+        );
+
+    let validate_invocation =
+        result.validate_call_info.as_ref().map(function_invocation_from_call_info);
+
+    let fee_transfer_invocation =
+        result.fee_transfer_call_info.as_ref().map(function_invocation_from_call_info);
+
+    let transaction_trace = match &transaction.transaction {
+        ExecutableTx::Declare(_) => TransactionTrace::Declare(DeclareTransactionTrace {
+            validate_invocation,
+            fee_transfer_invocation,
+            state_diff: None,
+        }),
+        ExecutableTx::DeployAccount(_) => {
+            TransactionTrace::DeployAccount(DeployAccountTransactionTrace {
+                constructor_invocation: function_invocation?,
+                validate_invocation,
+                fee_transfer_invocation,
+                state_diff: None,
+            })
+        }
+        ExecutableTx::Invoke(_) => TransactionTrace::Invoke(InvokeTransactionTrace {
+            validate_invocation,
+            execute_invocation: if let Some(revert_reason) = result.revert_error {
+                ExecuteInvocation::Reverted(RevertedInvocation { revert_reason })
+            } else {
+                ExecuteInvocation::Success(function_invocation?)
+            },
+            fee_transfer_invocation,
+            state_diff: None,
+        }),
+        ExecutableTx::L1Handler(_) => TransactionTrace::L1Handler(L1HandlerTransactionTrace {
+            function_invocation: function_invocation?,
+            state_diff: None,
+        }),
+    };
+
+    let execute_gas_consumed =
+        result.execute_call_info.map(|e| e.execution.gas_consumed).unwrap_or_default();
+    let validate_gas_consumed =
+        result.validate_call_info.map(|e| e.execution.gas_consumed).unwrap_or_default();
+    let gas_consumed = execute_gas_consumed + validate_gas_consumed;
+    let overall_fee = result.actual_fee.0 as u64;
+    let gas_price = if gas_consumed != 0 { overall_fee / gas_consumed } else { 0 };
```

Comment:
why not just use the `calculate_execution_fee()`? tho, its a slightly wrong bcs the fee type is set to be wei by default so have to fix that.